extensions/smallrye-fault-tolerance/runtime/src/main/java/io/quarkus/smallrye/faulttolerance/runtime/QuarkusFaultToleranceOperationProvider.java

```diff
@@ -1,22 +1,50 @@
 package io.quarkus.smallrye.faulttolerance.runtime;
 
 import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.function.Function;
 
 import javax.annotation.Priority;
-import javax.enterprise.context.Dependent;
 import javax.enterprise.inject.Alternative;
+import javax.inject.Singleton;
+
+import org.jboss.logging.Logger;
 
 import io.smallrye.faulttolerance.FaultToleranceOperationProvider;
 import io.smallrye.faulttolerance.config.FaultToleranceOperation;
 
-@Dependent
+@Singleton
 @Alternative
 @Priority(1)
 public class QuarkusFaultToleranceOperationProvider implements FaultToleranceOperationProvider {
+    private static final Logger LOG = Logger.getLogger(QuarkusFaultToleranceOperationProvider.class);
+    private Map<String, FaultToleranceOperation> operationCache;
+
+    /**
+     * Called by SmallryeFaultToleranceRecorder to init the operation cache.
+     */
+    void init(Map<String, FaultToleranceOperation> operationCache) {
+        this.operationCache = operationCache;
+    }
 
     @Override
     public FaultToleranceOperation get(Class<?> beanClass, Method method) {
+        String key = getCacheKey(beanClass, method);
+        return operationCache.computeIfAbsent(key, new Function<String, FaultToleranceOperation>() {
+            @Override
+            public FaultToleranceOperation apply(String key) {
+                return createAtRuntime(beanClass, method, key);
+            }
+        });
+    }
+
+    private FaultToleranceOperation createAtRuntime(Class<?> beanClass, Method method, String key) {
+        LOG.warnf("FaultToleranceOperation not found in the cache for %s creating it at runtime", key);
```

Comment:
Yes, now that we define that it can occurs and it's not a bug, INFO seems better.

Reply:
Do we really need an INFO if it's the normal behavior? Or it's not a normal behavior?

If it's normal behavior, it should just be debug.

Reply:
It will be logged only once, and should arrives in very rare conditions (using the builder aproach for RestClient is one that we thin of, this aproach is not documented on the Quarkus site).
But I'm OK with DEBUG ...

Reply:
If it's something that can happen and doesn't need any action by the user then debug is enough IMHO.